src/marshmallow/schema.py

```diff
@@ -108,10 +113,20 @@ def __new__(mcs, name, bases, attrs):
         klass = super().__new__(mcs, name, bases, attrs)
         inherited_fields = _get_fields_by_mro(klass, base.FieldABC, ordered=ordered)
 
+        # Instantiate the options class using class Meta
         meta = klass.Meta
-        # Set klass.opts in __new__ rather than __init__ so that it is accessible in
-        # get_declared_fields
-        klass.opts = klass.OPTIONS_CLASS(meta, ordered=ordered)
+        if combine_opts and len(bases) > 1:
+            # Create a new options class that combines all the base classes
+            combined_opts = type(
+                name + "Opts", tuple({base.OPTIONS_CLASS for base in bases}), {}
+            )
+            # Instantiate the options class as we normally do
+            klass.opts = combined_opts(meta, ordered=ordered)
+        else:
+            # Set klass.opts in __new__ rather than __init__ so that it is accessible in
+            # get_declared_fields
+            klass.opts = klass.OPTIONS_CLASS(meta, ordered=ordered)
+
```

Comment:
Rewrite proposal:

```py
        if combine_opts and len(bases) > 1:
            # Create a new options class that combines all the base classes
            opts = type(
                name + "Opts", tuple({base.OPTIONS_CLASS for base in bases}), {}
            )
        else:
            opts = klass.OPTIONS_CLASS
       # Set klass.opts in __new__ rather than __init__ so that it is accessible in
       # get_declared_fields
        klass.opts = opts(meta, ordered=ordered)
```

Reply:
True, that is admittedly a lot less redundant. I'll fix it at some point